File: ios/App/Sources/Core/Domain/NavCaddy/Routing/RoutingOrchestrator.swift
import Foundation

/// Orchestrates routing decisions based on classified intents.
///
/// Spec R3: Maps intent -> module route + parameters, validates prerequisites,
/// executes navigation as deep link or returns no-navigation result.
///
/// Thread-safe actor that processes classification results and determines
/// the appropriate routing action.
actor RoutingOrchestrator {
    private let prerequisiteChecker: PrerequisiteChecker

    init(prerequisiteChecker: PrerequisiteChecker) {
        self.prerequisiteChecker = prerequisiteChecker
    }

    /// Processes a classification result and returns the appropriate routing action
    ///
    /// - Parameter result: The classification result from IntentClassifier
    /// - Returns: A RoutingResult indicating the action to take
    func route(_ result: ClassificationResult) async -> RoutingResult {
        switch result {
        case .route(let intent, let target):
            // High confidence: validate prerequisites and route
            return await routeWithPrerequisites(intent: intent, target: target)

        case .confirm(let intent, let message):
            // Medium confidence: need user confirmation
            return .confirmationRequired(intent: intent, message: message)

        case .clarify(let response):
            // Low confidence: return as no-navigation with clarification message
            // Note: ClarificationResponse should be handled by UI layer
            return .noNavigation(
                intent: ParsedIntent(
                    intentType: .helpRequest,
                    confidence: 0.0
                ),
                response: response.message
            )

        case .error(let error):
            // Classification error: return as no-navigation with error message
            return .noNavigation(
                intent: ParsedIntent(
                    intentType: .helpRequest,
                    confidence: 0.0
                ),
                response: "Sorry, I couldn't understand that. \(error.localizedDescription)"
            )
        }
    }

    /// Routes an intent after validating prerequisites
    private func routeWithPrerequisites(intent: ParsedIntent, target: RoutingTarget) async -> RoutingResult {
        // Check if this intent type requires navigation
        if isNoNavigationIntent(intent.intentType) {
            return .noNavigation(
                intent: intent,
                response: generateNoNavigationResponse(for: intent)
            )
        }

        // Get required prerequisites for this intent
        let requiredPrerequisites = getRequiredPrerequisites(for: intent.intentType)

        // Check all prerequisites
        var missingPrerequisites: [Prerequisite] = []
        for prerequisite in requiredPrerequisites {
            let satisfied = await prerequisiteChecker.check(prerequisite)
            if !satisfied {
                missingPrerequisites.append(prerequisite)
            }
        }

        // If any prerequisites are missing, return prerequisite missing result
        if !missingPrerequisites.isEmpty {
            let message = buildPrerequisiteMessage(
                intent: intent.intentType,
                missing: missingPrerequisites
            )
            return .prerequisiteMissing(
                intent: intent,
                missing: missingPrerequisites,
                message: message
            )
        }

        // All prerequisites satisfied, proceed with navigation
        return .navigate(target: target, intent: intent)
    }

    /// Determines if an intent type should not navigate
    private func isNoNavigationIntent(_ intentType: IntentType) -> Bool {
        switch intentType {
        case .patternQuery, .helpRequest, .feedback:
            return true
        default:
            return false
        }
    }

    /// Gets the required prerequisites for an intent type
    private func getRequiredPrerequisites(for intentType: IntentType) -> [Prerequisite] {
        switch intentType {
        case .recoveryCheck:
            return [.recoveryData]

        case .scoreEntry, .roundEnd:
            return [.roundActive]

        case .clubAdjustment, .shotRecommendation:
            return [.bagConfigured]

        case .courseInfo:
            return [.courseSelected]

        case .drillRequest:
            // Optionally requires bag configured, but can work without it
            return []

        case .weatherCheck, .statsLookup, .roundStart, .equipmentInfo,
             .settingsChange, .helpRequest, .feedback, .patternQuery:
            return []
        }
    }

    /// Builds a user-facing message for missing prerequisites
    private func buildPrerequisiteMessage(
        intent: IntentType,
        missing: [Prerequisite]
    ) -> String {
        if missing.count == 1 {
            return missing[0].missingMessage
        } else {
            let items = missing.map { $0.description }.joined(separator: ", ")
            return "You need to set up: \(items) first."
        }
    }

    /// Generates a no-navigation response for informational intents (Bones persona)
    private func generateNoNavigationResponse(for intent: ParsedIntent) -> String {
        switch intent.intentType {
        case .patternQuery:
            return "Let me check your miss patterns. Based on your recent shots, I'll give you insights."

        case .helpRequest:
            return "I'm Bones, your digital caddy. Ask me about club selection, check your recovery, enter scores, or get coaching tips. What can I help you with?"

        case .feedback:
            return "Thanks for the feedback! I'm always learning to serve you better."

        default:
            return "I understand. Let me help you with that."
        }
    }
}

File: android/app/src/main/java/com/example/app/domain/navcaddy/routing/RoutingOrchestrator.kt
package caddypro.domain.navcaddy.routing

import caddypro.domain.navcaddy.classifier.ClassificationResult
import caddypro.domain.navcaddy.models.IntentType
import caddypro.domain.navcaddy.models.ParsedIntent
import caddypro.domain.navcaddy.models.RoutingTarget
import javax.inject.Inject

/**
 * Orchestrates routing decisions based on classification results.
 *
 * Takes ClassificationResult from IntentClassifier and:
 * 1. Validates prerequisites based on intent type
 * 2. Determines if intent requires navigation or can be answered inline
 * 3. Returns appropriate RoutingResult
 *
 * Spec reference: navcaddy-engine.md R3, navcaddy-engine-plan.md Task 10
 */
class RoutingOrchestrator @Inject constructor(
    private val prerequisiteChecker: PrerequisiteChecker
) {
    /**
     * Process classification result into routing decision.
     *
     * @param classificationResult Result from intent classifier
     * @return Routing decision (Navigate, NoNavigation, PrerequisiteMissing, or ConfirmationRequired)
     */
    suspend fun route(classificationResult: ClassificationResult): RoutingResult {
        return when (classificationResult) {
            is ClassificationResult.Route -> {
                handleRouteClassification(classificationResult.intent, classificationResult.target)
            }

            is ClassificationResult.Confirm -> {
                // Mid-confidence intents need confirmation before routing
                RoutingResult.ConfirmationRequired(
                    intent = classificationResult.intent,
                    message = classificationResult.message
                )
            }

            is ClassificationResult.Clarify -> {
                // Clarification is handled by UI layer, but if we receive it here,
                // we should not route anywhere
                RoutingResult.NoNavigation(
                    intent = createPlaceholderIntent(classificationResult.originalInput),
                    response = classificationResult.message
                )
            }

            is ClassificationResult.Error -> {
                // Error is handled by UI layer
                RoutingResult.NoNavigation(
                    intent = createPlaceholderIntent("error"),
                    response = classificationResult.message
                )
            }
        }
    }

    /**
     * Handle high-confidence route classification.
     * Check prerequisites and determine final routing.
     */
    private suspend fun handleRouteClassification(
        intent: ParsedIntent,
        target: RoutingTarget
    ): RoutingResult {
        // Check if this is a no-navigation intent
        if (isNoNavigationIntent(intent.intentType)) {
            return RoutingResult.NoNavigation(
                intent = intent,
                response = generateNoNavigationResponse(intent)
            )
        }

        // Check prerequisites for intents that require data
        val requiredPrerequisites = getRequiredPrerequisites(intent.intentType)
        if (requiredPrerequisites.isNotEmpty()) {
            val missingPrerequisites = prerequisiteChecker.checkAll(requiredPrerequisites)
            if (missingPrerequisites.isNotEmpty()) {
                return RoutingResult.PrerequisiteMissing(
                    intent = intent,
                    missing = missingPrerequisites,
                    message = generatePrerequisiteMessage(intent.intentType, missingPrerequisites)
                )
            }
        }

        // All checks passed, navigate to target
        return RoutingResult.Navigate(
            target = target,
            intent = intent
        )
    }

    /**
     * Determine if intent type should not navigate to a new screen.
     */
    private fun isNoNavigationIntent(intentType: IntentType): Boolean {
        return when (intentType) {
            IntentType.PATTERN_QUERY,
            IntentType.HELP_REQUEST,
            IntentType.FEEDBACK -> true
            else -> false
        }
    }

    /**
     * Get required prerequisites for an intent type.
     */
    private fun getRequiredPrerequisites(intentType: IntentType): List<Prerequisite> {
        return when (intentType) {
            IntentType.RECOVERY_CHECK -> listOf(Prerequisite.RECOVERY_DATA)

            IntentType.SCORE_ENTRY,
            IntentType.ROUND_END -> listOf(Prerequisite.ROUND_ACTIVE)

            IntentType.CLUB_ADJUSTMENT -> listOf(Prerequisite.BAG_CONFIGURED)

            IntentType.SHOT_RECOMMENDATION -> listOf(Prerequisite.BAG_CONFIGURED)

            IntentType.COURSE_INFO -> listOf(Prerequisite.COURSE_SELECTED)

            else -> emptyList()
        }
    }

    /**
     * Generate user-facing message for missing prerequisites.
     */
    private fun generatePrerequisiteMessage(
        intentType: IntentType,
        missingPrerequisites: List<Prerequisite>
    ): String {
        return when {
            missingPrerequisites.contains(Prerequisite.RECOVERY_DATA) -> {
                "I don't have any recovery data yet. Log your sleep, HRV, or readiness score first, and I'll give you insights."
            }

            missingPrerequisites.contains(Prerequisite.ROUND_ACTIVE) -> {
                "You need to start a round first. Would you like to start a new round now?"
            }

            missingPrerequisites.contains(Prerequisite.BAG_CONFIGURED) -> {
                "Your bag isn't configured yet. Set up your clubs and distances so I can give you better recommendations."
            }

            missingPrerequisites.contains(Prerequisite.COURSE_SELECTED) -> {
                "Which course are you playing? Select a course to get specific information."
            }

            else -> {
                "Some required information is missing. Please complete your profile first."
            }
        }
    }

    /**
     * Generate response for no-navigation intents.
     */
    private fun generateNoNavigationResponse(intent: ParsedIntent): String {
        return when (intent.intentType) {
            IntentType.PATTERN_QUERY -> {
                "Let me check your miss patterns. Based on your recent shots, I'll give you insights."
            }

            IntentType.HELP_REQUEST -> {
                "I'm Bones, your digital caddy. Ask me about club selection, check your recovery, " +
                        "enter scores, or get coaching tips. What can I help you with?"
            }

            IntentType.FEEDBACK -> {
                "Thanks for the feedback! I'm always learning to serve you better."
            }

            else -> {
                "I understand. Let me help you with that."
            }
        }
    }

    /**
     * Create placeholder intent for non-intent results (clarify/error).
     */
    private fun createPlaceholderIntent(input: String): ParsedIntent {
        return ParsedIntent(
            intentId = "placeholder_${System.currentTimeMillis()}",
            intentType = IntentType.HELP_REQUEST,
            confidence = 0f,
            entities = caddypro.domain.navcaddy.models.ExtractedEntities(),
            userGoal = input,
            routingTarget = null
        )
    }
}
